AdblockPlusSafari/AdblockPlusExtras.m

```diff
@@ -0,0 +1,269 @@
+    [self.downloadTasks removeObjectForKey:filterlistName];
+  }
+}
+
+- (void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didFinishDownloadingToURL:(NSURL *)location
+{
+  NSString *filterlistName = downloadTask.originalRequest.URL.absoluteString;
+  NSDictionary *filterlist = self.filterlists[filterlistName];
+
+  if ([filterlist[@"taskIdentifier"] unsignedIntegerValue] == downloadTask.taskIdentifier) {
+
+    NSString *errorMessage;
+
+    {
+      if (![downloadTask.response isKindOfClass:[NSHTTPURLResponse class]]) {
+        // This error occurs in rare cases. The error message is meaningless to ordinary user.
+        NSLog(@"Downloading has failed: %@", downloadTask.error);
+        errorMessage = @"";
+        goto END;
+      }
+
+      NSHTTPURLResponse *response = (NSHTTPURLResponse *)downloadTask.response;
+
+      if (response.statusCode < 200 || response.statusCode >= 300) {
+        errorMessage = [NSString stringWithFormat:@" Remote server responded: %ld (%@).", response.statusCode, [NSHTTPURLResponse localizedStringForStatusCode:response.statusCode]];
+        goto END;
+      }
+
+      NSFileManager *fileManager = [[NSFileManager alloc] init];
+      fileManager.delegate = self;
+
+      // http://www.atomicbird.com/blog/sharing-with-app-extensions
+      NSURL *destination = [fileManager containerURLForSecurityApplicationGroupIdentifier:self.group];
+      destination = [destination URLByAppendingPathComponent:filterlist[@"filename"] isDirectory:NO];
+
+      NSError *error;
+      // http://stackoverflow.com/questions/20683696/how-to-overwrite-a-folder-using-nsfilemanager-defaultmanager-when-copying
+      if (![fileManager moveItemAtURL:location toURL:destination error:&error]) {
+        NSLog(@"Moving has failed: %@", error);
+        errorMessage = @"";
+        goto END;
+      }
+
+      // Success, store the result
+      NSMutableDictionary *mutableFilterlist = [filterlist mutableCopy];
+      mutableFilterlist[@"updating"] = @NO;
+      mutableFilterlist[@"lastUpdate"] = [NSDate date];
+      mutableFilterlist[@"downloaded"] = @YES;
+      self.downloadedVersion += 1;
+
+      NSMutableDictionary *mutableFilterlists = [self.filterlists mutableCopy];
+      mutableFilterlists[filterlistName] = mutableFilterlist;
+      self.filterlists = mutableFilterlists;
+    }
+
+  END:
+    if (errorMessage) {
+      UIViewController *viewController = UIApplication.sharedApplication.delegate.window.rootViewController;
+      while (viewController.presentedViewController) {
+        viewController = viewController.presentedViewController;
+      }
+
+      NSString *title = @"Updating filterlists";
+      NSString *message = [NSString stringWithFormat:@"Filterlist %@ cannot be updated.%@", filterlistName, errorMessage];
+
+      UIAlertController *alertController = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
```

Comment:
Correct, it is not re-attempted. It will try it again with next refresh. The problem is that iOS allow apps execute in background for just limited time.